bittide-instances/src/Bittide/Instances/Hitl/Driver/VexRiscv.hs

```diff
@@ -48,38 +48,38 @@ driverFunc _name targets = do
 
   let
     catchError :: DeviceInfo -> SomeException -> IO ExitCode
-    catchError d ex = do
-      putStrLn $ "Test Failed on device " <> d.deviceId <> " with: " <> displayException ex
+    catchError dI ex = do
+      putStrLn $ "Test Failed on device " <> dI.deviceId <> " with: " <> displayException ex
       pure $ ExitFailure 2
 
-  exitCodes <- forM (L.zip [0 ..] targets) $ \(targetIndex, (hwT, d)) -> handle (liftIO . catchError d) $ do
-    liftIO $ putStrLn $ "Running driver for " <> d.deviceId
+  projectDir <- liftIO $ findParentContaining "cabal.project"
+
+  exitCodes <- forM (L.zip [0 ..] targets) $ \(targetIndex, (hwT, dI)) -> handle (liftIO . catchError dI) $ do
+    liftIO $ putStrLn $ "Running driver for " <> dI.deviceId
+
+    let
+      hitlDir = projectDir </> "_build" </> "hitl"
+      mkLogPath str = (hitlDir </> str <> show targetIndex <> ".log")
+      picoOutLog = mkLogPath "picocom-stdout"
+      picoErrLog = mkLogPath "picocom-stderr"
+      ocdOutLog = mkLogPath "openocd-stdout"
+      ocdErrLog = mkLogPath "openocd-stderr"
+      openocdEnv = [("OPENOCD_STDOUT_LOG", ocdOutLog), ("OPENOCD_STDERR_LOG", ocdErrLog)]
+      gdbPort = 3333 + targetIndex :: Integer
 
     openHardwareTarget hwT
 
     -- even though this is just pre-process step, the CPU is reset until
     -- the test_start signal is asserted and cannot be accessed via GDB otherwise
     updateVio "vioHitlt" [("probe_test_start", "1")]
-
-    let gdbPort = 3333 + targetIndex
-    liftIO $ withOpenOcd d.usbAdapterLocation gdbPort 6666 4444 $ \ocd -> do
+    liftIO $ withOpenOcdWithEnv openocdEnv dI.usbAdapterLocation 3333 6666 4444 $ \ocd -> do
```

Comment:
Shouldn't the `3333` be `gdbPort` like before?
So it can be made unique by the `+ targetIndex` part?
And it'll match what you later feed to `runGdbCommands`.

Reply:
Thanks for noticing :)